content/blog/2022-05-24-local-experiments-to-cloud-with-tpi-docker.md

```diff
@@ -0,0 +1,132 @@
+the software versions, and also makes it trivial to switch between cloud
+providers -- currently Amazon Web Services (AWS), Microsoft Azure, Google Cloud
+Platform, and Kubernetes. Your Docker image is cloud provider-agnostic. There
+are thousands of [pre-defined Docker images online](https://hub.docker.com/)
+too.
+
+In this tutorial, we'll use an existing Docker image that comes with most of our
+requirements already installed. We'll then add add a few more dependencies on
+top and run our training pipeline in the cloud as before!
+
+## Run GPU-enabled Docker containers
+
+<admon type="warn">
+
+If you haven't read the [previous tutorial][bees-part-1], you should check out
+the basics there first. This includes how to let Terraform know about TPI, and
+essential commands (`init`, `apply`, `refresh`, `show`, and `destroy`).
+
+</admon>
+
+The only modification from the [previous tutorial][bees-part-1] is the script
+part of the `main.tf` config file.
+
+Let's say we've found a carefully prepared a Docker image suitable for data
+science and machine learning -- in this case,
+[`iterativeai/cml:0-dvc2-base1-gpu`](https://cml.dev/doc/self-hosted-runners#docker-images).
+This image comes loaded with Ubuntu 20.04, Python 3.8, NodeJS, CUDA 11.0.3,
+CuDNN 8, Git, [CML](https://cml.dev), [DVC](https://dvc.org), and other
+essentials for full-stack data science.
+
+Our `script` block is now:
+
+```hcl
+script = <<-END
+  #!/bin/bash
+  docker run --gpus all -v "$PWD:/tpi" -w /tpi -e TF_CPP_MIN_LOG_LEVEL \
+    iterativeai/cml:0-dvc2-base1-gpu /bin/bash -c "
+  pip install -r requirements.txt tensorflow==2.8.0
+  python train.py --output results-gpu/metrics.json
+  "
+END
+```
+
+Yes, it's quite long for a one-liner. Let's looks at the components:
+
+- `docker run`: Download the specified image, create a container from the image,
+  and run it.
+- `--gpus all`: Expose GPUs to the container.
+- `-v "$PWD:/tpi"`: Expose our current working directory (`$PWD`) within the
+  container (as path `/tpi`).
+- `-w /tpi`: Set the working directory of the container (to be `/tpi`).
+- `-e TF_CPP_MIN_LOG_LEVEL`: Expose the environment variable
+  `TF_CPP_MIN_LOG_LEVEL` to the container (in this case to control TensorFlow's
+  verbosity).
+- `iterativeai/cml:0-dvc2-base1-gpu`: The image we want to download & run a
+  container from.
+- `/bin/bash -c "pip install -r requirements.txt ... python train.py ..."`:
+  Commands to run within the container's working directory. In this case,
+  install the dependencies and run the training script.
+
+We can now call `terraform init` and `terraform apply` to provision
+infrastructure, sync data and code to it, set up the environment, and run the
+training process. If you'd like to tinker with this example you can
+[find it on GitHub](https://github.com/iterative/blog-tpi-bees/tree/docker).
+
+<admon type="tip">
+
+Don't forget to `terraform refresh && terraform show` to check the status, and
```

Comment:
Worth mentioning a bit of the `TF_LOG_PROVIDER` awesomeness? 🦄🌈🤢

Reply:
`I`DK